### DISCUSSION TITLE
Ability to push builded VM appliance to S3, OCI registry, Openstack Glance Service,...

### DISCUSSION THREAD

**Discussion by tannevaled** [lima-vm/lima#2896] [Ideas]
It is common for the container ecosystem to pull/push.
With the nested virtualization availability, it would be very useful to benefit from the same workflow for VMs.

# Sample usage

lima.toml
```
[[s3]]
ENDPOINT=s3.example.com
AWS_ACCESS_KEY_ID=XXX
AWS_SECRET_ACCESS_KEY=YYY
BUCKET=mybucket

[[oci]]
ENDPOINT=harbor.example.com
LOGIN=XXX
PASSWORD=YYY

[[os]]
ENDPOINT=openstack.example.com
...
```

```bash
$ limactl push devstack --endpoint="s3://s3.example.com"
$ limactl push devstack --endpoint="oci://harbor.example.com" --tag="latest"
$ limactl push devstack --endpoint="os://openstack.example.com" --properties="DISTRIBUTION_NAME=ubuntu,DISTRIBUTION_VERSION=24.04" 
```

**Comment by nirs** [comment]
Pushing an image is not easy:
- Pushing raw sparse image is very slow and inefficient, unless the server supports sparseness. I think that no server supports this expect ovirt-imageio which part of ovirt, you probably don't want to push image to ovirt at this point.
- Pushing qcow2 compressed is best, but it depends on how do you want to use the image later. Do you plan to use to the image with lima? (it knows how to read qcow2 images).

You can use an existing tool to push the image form `~/.lima/name/diffdisk`. Why lima is a better tool for this?

**Reply by AkihiroSuda** [reply]
Pushing a VM image is quite challenging in terms of security, as the VM image contains data of removed files.

**Reply by tannevaled** [reply]
> 
>     * Pushing qcow2 compressed is best, but it depends on how do you want to use the image later. Do you plan to use to the image with lima? (it knows how to read qcow2 images).

i plan to use it mainly with OpenStack.

**Comment by tannevaled** [comment]
just to document a little https://tart.run/integrations/vm-management/#working-with-a-remote-oci-container-registry
